File: golang/sign-demo/SignTool.java
package com.dmall.smartimage.depot.controller;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * @author Fan Huaran
 * created on 2019/10/29
 * @description
 */
public class SignTool {

    private static byte[] CONSTANT_BYTES = { 0x65, 0x62, 0x72, 0x63, 0x55, 0x59, 0x69, 0x75, 0x78, 0x61, 0x5a, 0x76, 0x32, 0x58, 0x47, 0x75, 0x37, 0x4b, 0x49, 0x59, 0x4b, 0x78, 0x55, 0x72, 0x71, 0x66, 0x6e, 0x4f, 0x66, 0x70, 0x44, 0x46 };

    public static String sign(MySign mySign, String tempStr)
    {
        MessageDigest tempStrMessageDigest = getShaDigest();
        tempStrMessageDigest.update(tempStr.getBytes(StandardCharsets.UTF_8));
        byte[] tempStrSha1Bytes = tempStrMessageDigest.digest() ;

        byte[] tempStrSha1HexBytes = hexEncode4Bytes(tempStrSha1Bytes);
        byte[] uidBytes = String.valueOf(mySign.uid).getBytes(StandardCharsets.UTF_8);
        byte[] timeBytes = String.valueOf(mySign.time).getBytes(StandardCharsets.UTF_8);
        byte[] devUIDBytes = mySign.devUID.getBytes(StandardCharsets.UTF_8);

        byte[] mergeBytes = mergeBytes(tempStrSha1HexBytes, uidBytes, CONSTANT_BYTES, timeBytes, devUIDBytes);

        MessageDigest signMessageDigest = getShaDigest();
        signMessageDigest.update(mergeBytes);
        byte[] resultSha1Bytes = signMessageDigest.digest();
        return hexEncode(resultSha1Bytes);
    }

    public static String getDevUID(String feature)
    {
        return "O|" + md5(feature).toUpperCase();
    }

    private static MessageDigest getShaDigest(){
        try {
            return MessageDigest.getInstance("SHA");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private static String md5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            md.update(input.getBytes(StandardCharsets.UTF_8));
            byte[] outBytes = md.digest();
            return hexEncode(outBytes);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private static String hexEncode(byte[] input) {
        StringBuilder builder = new StringBuilder();
        for(byte value : input) {
            String hex = Integer.toHexString(value & 0xFF);
            if(hex.length() < 2){
                builder.append("0");
            }
            builder.append(hex);
        }
        return builder.toString();
    }

    private static byte[] hexEncode4Bytes(byte[] input) {
        return hexEncode(input).getBytes(StandardCharsets.UTF_8);
    }

    private static byte[] mergeBytes(byte[]... bytesArray) {
        int length = 0;
        for (byte[] value : bytesArray) {
            length += value.length;
        }

        byte[] mergeBytes = new byte[length];
        int index = 0;
        for (byte[] value : bytesArray) {
            for (int i = 0, n = value.length; i < n; i++) {
                mergeBytes[index++] = value[i];
            }
        }

        return mergeBytes;
    }

    /**
     * input
     */
    public static class MySign {

        private long uid;

        private String devUID;

        private long time;

        public long getUid() {
            return uid;
        }

        public void setUid(long uid) {
            this.uid = uid;
        }

        public String getDevUID() {
            return devUID;
        }

        public void setDevUID(String devUID) {
            this.devUID = devUID;
        }

        public long getTime() {
            return time;
        }

        public void setTime(long time) {
            this.time = time;
        }
    }

    public static void main(String[] args) {
        MySign mySign = new MySign();
        mySign.setUid(1L);
        mySign.setDevUID(SignTool.getDevUID("helloworld"));
        mySign.setTime(1L);
        System.out.println(sign(mySign,  "hello"));
    }

}

File: golang/sign-demo/SignTool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace APP
{
    /// <summary>
    /// 
    /// </summary>
    class MySign
    {
        public Int64 Uid { get; set; }
        public Int64 Time { get; set; }
        public String DevUID { get; set; }
    }

    class Helper
    {
        public static string Md5(string str)
        {
            MD5 md5 = MD5.Create();
            byte[] resultBytes = md5.ComputeHash(Encoding.UTF8.GetBytes(str));
            return HexEncode(resultBytes);
        }

        public static String HexEncode(byte[] input)
        {
            StringBuilder tmp = new StringBuilder();
            foreach (byte i in input)
            {
                tmp.Append(i.ToString("x2"));
            }
            return tmp.ToString();
        }

        public static byte[] HexEncode4Bytes(byte[] input)
        {
            return Encoding.UTF8.GetBytes(HexEncode(input));
        }

        public static byte[] MergeBytes(byte[][] bytesArray)
        {
            var length = 0;
            foreach (byte[] value in bytesArray)
            {
                length += value.Length;
            }
            byte[] mergeBytes = new byte[length];

            int index = 0;
            foreach (byte[] value in bytesArray)
            {
                value.CopyTo(mergeBytes, index);
                index += value.Length;
            }

            return mergeBytes;
        }
    }

    class SignTool
    {
        private static byte[] CONSTANT_BYTES = { 0x65, 0x62, 0x72, 0x63, 0x55, 0x59, 0x69, 0x75, 0x78, 0x61, 0x5a, 0x76, 0x32, 0x58, 0x47, 0x75, 0x37, 0x4b, 0x49, 0x59, 0x4b, 0x78, 0x55, 0x72, 0x71, 0x66, 0x6e, 0x4f, 0x66, 0x70, 0x44, 0x46 };

        public static String Sign(MySign mySign, string tempStr)
        {
            var sha1CryptoServiceProvider = new SHA1CryptoServiceProvider();
            byte[] tempStrSha1Bytes = sha1CryptoServiceProvider.ComputeHash(Encoding.UTF8.GetBytes(tempStr));

            byte[] tempStrSha1HexBytes = Helper.HexEncode4Bytes(tempStrSha1Bytes);
            byte[] uidBytes = Encoding.UTF8.GetBytes(mySign.Uid.ToString());
            byte[] timeBytes = Encoding.UTF8.GetBytes(mySign.Time.ToString());
            byte[] devUIDBytes = Encoding.UTF8.GetBytes(mySign.DevUID);

            byte[] mergeBytes = Helper.MergeBytes(new byte[][] { tempStrSha1HexBytes, uidBytes, CONSTANT_BYTES, timeBytes, devUIDBytes });

            byte[] resultSha1Bytes = sha1CryptoServiceProvider.ComputeHash(mergeBytes);
            return Helper.HexEncode(resultSha1Bytes);
        }

        public static string getDevUID(String feature)
        {
            return "O|" + Helper.Md5(feature).ToUpper();
        }
    }

    class APP
    {
        static void Main(string[] args)
        {
            //var rightValue = "O|FC5E038D38A57032085441E7FE7010B0";
            //var testValue = SignTool.getDevUID("helloworld");
            //Console.WriteLine(rightValue == testValue);
            //Console.WriteLine(testValue);

            MySign mySign = new MySign();
            mySign.Uid = 1L;
            mySign.DevUID = SignTool.getDevUID("helloworld");
            mySign.Time = 1L;
            var outSign = SignTool.Sign(mySign, "hello");
            Console.WriteLine(outSign);
            Console.ReadKey();
        }
    }
}
